diff --git a/qtbase/src/plugins/styles/windowsvista/qwindowsvistastyle.cpp b/qtbase/src/plugins/styles/windowsvista/qwindowsvistastyle.cpp
index 4967ac3afa4b95e0cbabcec75a8bd6229a5d3b05..3d20358bfd57c44f77333d88cb1be48892fbec7c 100644
--- a/qtbase/src/plugins/styles/windowsvista/qwindowsvistastyle.cpp
+++ b/qtbase/src/plugins/styles/windowsvista/qwindowsvistastyle.cpp
@@ -1193,9 +1193,11 @@ void QWindowsVistaStyle::drawControl(ControlElement element, const QStyleOption
             QPoint p1 = QStyle::visualPos(option->direction, menuitem->rect, QPoint(checkcol, rect.top()));
             QPoint p2 = QStyle::visualPos(option->direction, menuitem->rect, QPoint(checkcol, rect.bottom()));
             QRect gutterRect(p1.x(), p1.y(), gutterWidth, p2.y() - p1.y() + 1);
+/*
             XPThemeData theme2(widget, painter, QWindowsXPStylePrivate::MenuTheme,
                                MENU_POPUPGUTTER, stateId, gutterRect);
             d->drawBackground(theme2);
+*/
 
             int x, y, w, h;
             menuitem->rect.getRect(&x, &y, &w, &h);
