diff --git forkSrcPrefix/qtwebengine/configure.cmake forkDstPrefix/qtwebengine/configure.cmake
index 5efb3d00daba58eb03895a7cd17d6c7eee7918f1..ddfbbd7001d4c5fcad0c17bef6f8464fcb27ec30 100644
--- forkSrcPrefix/qtwebengine/configure.cmake
+++ forkDstPrefix/qtwebengine/configure.cmake
@@ -49,10 +49,6 @@ if(PkgConfig_FOUND)
 endif()
 
 #### Tests
-if(LINUX)
-   check_for_ulimit()
-endif()
-
 qt_config_compile_test(re2
     LABEL "re2"
     LIBRARIES
