diff --git a/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp b/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp
index 48b9bc5eba7c6885d8f499b4961df04d10c95556..b40ba5354e31255a442666923952d4ba6fcda0ae 100644
--- a/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp
+++ b/qtbase/src/plugins/platforms/windows/uiautomation/qwindowsuiamainprovider.cpp
@@ -214,12 +214,16 @@ void QWindowsUiaMainProvider::notifyValueChange(QAccessibleValueChangeEvent *eve
 void QWindowsUiaMainProvider::notifyNameChange(QAccessibleEvent *event)
 {
     if (QAccessibleInterface *accessible = event->accessibleInterface()) {
-        if (QWindowsUiaMainProvider *provider = providerForAccessible(accessible)) {
-            VARIANT oldVal, newVal;
-            clearVariant(&oldVal);
-            setVariantString(accessible->text(QAccessible::Name), &newVal);
-            QWindowsUiaWrapper::instance()->raiseAutomationPropertyChangedEvent(provider, UIA_NamePropertyId, oldVal, newVal);
-            ::SysFreeString(newVal.bstrVal);
+        // Restrict notification to combo boxes, which need it for accessibility,
+        // in order to avoid slowdowns with unnecessary notifications.
+        if (accessible->role() == QAccessible::ComboBox) {
+            if (QWindowsUiaMainProvider *provider = providerForAccessible(accessible)) {
+                VARIANT oldVal, newVal;
+                clearVariant(&oldVal);
+                setVariantString(accessible->text(QAccessible::Name), &newVal);
+                QWindowsUiaWrapper::instance()->raiseAutomationPropertyChangedEvent(provider, UIA_NamePropertyId, oldVal, newVal);
+                ::SysFreeString(newVal.bstrVal);
+            }
         }
     }
 }
