diff --git forkSrcPrefix/qtbase/src/widgets/accessible/complexwidgets.cpp forkDstPrefix/qtbase/src/widgets/accessible/complexwidgets.cpp
index 88bafee82f480977586af4f1df2f031d2ec16adb..1832f5f46ee83703a01d3b5c67a245b82942c579 100644
--- forkSrcPrefix/qtbase/src/widgets/accessible/complexwidgets.cpp
+++ forkDstPrefix/qtbase/src/widgets/accessible/complexwidgets.cpp
@@ -490,13 +490,19 @@ QWidgetList QAccessibleAbstractScrollArea::accessibleChildren() const
     // Horizontal scrollBar container.
     QScrollBar *horizontalScrollBar = abstractScrollArea()->horizontalScrollBar();
     if (horizontalScrollBar && horizontalScrollBar->isVisible()) {
-        children.append(horizontalScrollBar->parentWidget());
+        QWidget *scrollBarParent = horizontalScrollBar->parentWidget();
+        // Add container only if scroll bar is in the container
+        if (elementType(scrollBarParent) == HorizontalContainer)
+            children.append(scrollBarParent);
     }
 
     // Vertical scrollBar container.
     QScrollBar *verticalScrollBar = abstractScrollArea()->verticalScrollBar();
     if (verticalScrollBar && verticalScrollBar->isVisible()) {
-        children.append(verticalScrollBar->parentWidget());
+        QWidget *scrollBarParent = verticalScrollBar->parentWidget();
+        // Add container only if scroll bar is in the container
+        if (elementType(scrollBarParent) == VerticalContainer)
+            children.append(scrollBarParent);
     }
 
     // CornerWidget.
