diff --git a/qtbase/src/widgets/kernel/qwidget.cpp b/qtbase/src/widgets/kernel/qwidget.cpp
index 71173b06e0b9e5535bc65097f9d3d26c84096aed..9787ba56d5544cd9c42a30583c4441c506d2ba0f 100644
--- a/qtbase/src/widgets/kernel/qwidget.cpp
+++ b/qtbase/src/widgets/kernel/qwidget.cpp
@@ -12585,15 +12585,7 @@ void QWidget::activateWindow()
 */
 int QWidget::metric(PaintDeviceMetric m) const
 {
-    QWindow *topLevelWindow = nullptr;
-    QScreen *screen = nullptr;
-    if (QWidget *topLevel = window()) {
-        topLevelWindow = topLevel->windowHandle();
-        if (topLevelWindow)
-            screen = topLevelWindow->screen();
-    }
-    if (!screen && QGuiApplication::primaryScreen())
-        screen = QGuiApplication::primaryScreen();
+    QScreen *screen = this->screen();
 
     if (!screen) {
         if (m == PdmDpiX || m == PdmDpiY)
@@ -12628,12 +12620,11 @@ int QWidget::metric(PaintDeviceMetric m) const
     } else if (m == PdmPhysicalDpiY) {
         return qRound(screen->physicalDotsPerInchY());
     } else if (m == PdmDevicePixelRatio) {
-        return topLevelWindow ? topLevelWindow->devicePixelRatio() : qApp->devicePixelRatio();
+        return screen->devicePixelRatio();
     } else if (m == PdmDevicePixelRatioScaled) {
-        return (QPaintDevice::devicePixelRatioFScale() *
-                (topLevelWindow ? topLevelWindow->devicePixelRatio() : qApp->devicePixelRatio()));
+        return QPaintDevice::devicePixelRatioFScale() * screen->devicePixelRatio();
     } else {
-        val = QPaintDevice::metric(m);// XXX
+        val = QPaintDevice::metric(m);
     }
     return val;
 }
