diff --git forkSrcPrefix/qtbase/src/plugins/platforms/windows/qwindowsscreen.cpp forkDstPrefix/qtbase/src/plugins/platforms/windows/qwindowsscreen.cpp
index 22be39bd921e9d591e172cd1537fa7c441c4ef6c..b5acf36169c0f9525d047b207589506502d6bc6b 100644
--- forkSrcPrefix/qtbase/src/plugins/platforms/windows/qwindowsscreen.cpp
+++ forkDstPrefix/qtbase/src/plugins/platforms/windows/qwindowsscreen.cpp
@@ -549,19 +549,19 @@ bool QWindowsScreenManager::isSingleScreen()
 }
 
 static inline int indexOfMonitor(const QWindowsScreenManager::WindowsScreenList &screens,
-                                 const QString &monitorName)
+                                 const QString &serialNumberIn)
 {
-    for (int i= 0; i < screens.size(); ++i)
-        if (screens.at(i)->data().name == monitorName)
+    for (int i = 0; i < screens.size(); ++i)
+        if (screens.at(i)->data().serialNumber == serialNumberIn)
             return i;
     return -1;
 }
 
 static inline int indexOfMonitor(const WindowsScreenDataList &screenData,
-                                 const QString &monitorName)
+                                 const QString &serialNumberIn)
 {
     for (int i = 0; i < screenData.size(); ++i)
-        if (screenData.at(i).name == monitorName)
+        if (screenData.at(i).serialNumber == serialNumberIn)
             return i;
     return -1;
 }
@@ -627,7 +627,7 @@ bool QWindowsScreenManager::handleScreenChanges()
     const bool lockScreen = newDataList.size() == 1 && (newDataList.front().flags & QWindowsScreenData::LockScreen);
     bool primaryScreenChanged = false;
     for (const QWindowsScreenData &newData : newDataList) {
-        const int existingIndex = indexOfMonitor(m_screens, newData.name);
+        const int existingIndex = indexOfMonitor(m_screens, newData.serialNumber);
         if (existingIndex != -1) {
             m_screens.at(existingIndex)->handleChanges(newData);
             if (existingIndex == 0)
@@ -644,7 +644,7 @@ bool QWindowsScreenManager::handleScreenChanges()
     // temporary lock screen to avoid window recreation (QTBUG-33062).
     if (!lockScreen) {
         for (int i = m_screens.size() - 1; i >= 0; --i) {
-            if (indexOfMonitor(newDataList, m_screens.at(i)->data().name) == -1)
+            if (indexOfMonitor(newDataList, m_screens.at(i)->data().serialNumber) == -1)
                 removeScreen(i);
         }     // for existing screens
     }     // not lock screen
