diff --git a/qtbase/src/gui/configure.cmake b/qtbase/src/gui/configure.cmake
index ffa4a4dce79e57aeaff2ca600ebb6ccdc62ff7aa..4d98a1a2eff26f1c0f5c8238ade5f14b03d58033 100644
--- a/qtbase/src/gui/configure.cmake
+++ b/qtbase/src/gui/configure.cmake
@@ -31,7 +31,6 @@ qt_find_package(DirectFB PROVIDED_TARGETS PkgConfig::DirectFB MODULE_NAME gui QM
 qt_find_package(Libdrm PROVIDED_TARGETS Libdrm::Libdrm MODULE_NAME gui QMAKE_LIB drm)
 qt_find_package(EGL PROVIDED_TARGETS EGL::EGL MODULE_NAME gui QMAKE_LIB egl)
 qt_find_package(WrapSystemFreetype 2.2.0 PROVIDED_TARGETS WrapSystemFreetype::WrapSystemFreetype MODULE_NAME gui QMAKE_LIB freetype)
-set_package_properties(WrapFreetype PROPERTIES TYPE REQUIRED)
 if(QT_FEATURE_system_zlib)
     qt_add_qmake_lib_dependency(freetype zlib)
 endif()
